### ISSUE TITLE
Improve PATHMAP safeguards

### ISSUE THREAD

**Issue by j26151** [ZeligsoftDev/CX4CBDDS#191]
# **Issue and tracking information**

### Developer's Actual time spent on fix (hours)  

# **Issue reporter to provide a detailed description of the issue in the space below**
When using the pathmap feature, we request that the following safeguards/measures be investigated to improve our user's ease of use with the feature:

- It has been noted in the past that when model references are broken, and models are imported, or imported with out of date models, that the model attempts to auto fix problems without user intervention. We would like to verify/ensure that there is no automatic fixing of model issues without the user agreeing first.
- When a referenced model is missing, throw a warning dialog to the user. This can be done when they open the dependent model, or when a model is deleted, making an open dependent model invalid
- When there are multiple models with the same model name and PATHMAP, throw a warning dialog to the user with instructions stating what model has been "chosen", how to select different paths to models, and that it is generally discouraged to continue

**Comment by ysroh** [comment]
Following safeguards are added for dynamic pathmap:

- Warn users if the dynamic model library is removed and there are dependent models in the workspace.
- Warn users if two models with the same dynamic pathmap resolve to different locations. In this case, users are notified that the target location will switch from an old to a new location. Warn users that keeping more than one model with the same dynamic pathmap can cause the dependent models to be inconsistent.
- Papyrus does not automatically fix any broken references without users' consent so we don't have to do anything for that. However, we may need to fix Issue #238 to show the broken references from the properties view.

**Comment by ysroh** [comment]
I found some issues with this fix so I will be committing some fixes as part of Issue #241.

**Comment by j26151** [comment]
Can you let me know when you think this issue will be resolved on your end so that I can build and test it on mine? Thank you

**Comment by ysroh** [comment]
@j26151 Sure. I will let you know when it is ready for the test. Thanks

**Comment by ysroh** [comment]
@j26151 Can you please test this feature on your environment and let me know if everything is ok?

Update:
1. Warn users about dependent models to the pathmap library deleted. It will prompt users to close editors for all dependent models since Papyrus does the same thing when referenced model is removed from the workspace. The reason is that the removed pathmap library may still exist in the resource set of dependent model until the model is closed and reopened. 
2. Warn users about conflict dynamic pathmap and tell users which pathmap will be used.

**Comment by j26151** [comment]
I've noticed 2 issues.  Lets have a simple situation; There is a Model A that depends on a Model B using pathmap.

-  The first issue I believe is just not implemented. If Model B does not exist in the workspace, opening Model A should throw a warning to the user to let them know that they are opening a model that is missing a pathmap dependency and also name the missing dependencies so that they don't have to guess which one they are missing.
- The second issue I noticed is with the deletion of of Model B. I receive no warning at all about Model B being deleted at first. It takes some combination of opening Model A, validating, deleting Model B, and restarting Papyrus to get Papyrus to "understand" that Model B is has a Pathmap and Model A depends on it. I believe at a minimum I had to delete Model B from the workspace, restart Papyrus, open Model A (we would want to see the warning mentioned in the first bullet), close Model A, re-import Model B, open Model A (verify everything looks good), either close or leave the model A open, and remove model B from papyrus.  Actually, talking through it right now, it almost appears that Papyrus doesn't recognize a model as a Pathmap model in regards to throwing the deletion warning unless that model has been imported. In other words, if I bring in new Models A and B with no pathmapping, set up the pathmap on model B, import it to model A, then delete model B, Papyrus won't say anything because Model B has not been imported yet. I would need to import model B and then delete it to get the warning.

I know the second bullet trails on a bit and we can discuss this more if you like..

**Comment by j26151** [comment]
> ys] I would think that users would run validation to catch any broken links like any other issues in the model but I will look into it to see if I can add a listener when the model is open.

The problem is that we are going to have users bringing a bunch of models, since we have plenty of ICMs. Then they will open the one that they want and start making changes well before ever validating. They might see ports that are empty (references don't exist and they show the broken references) and this will confuse them. It would be good to avoid this if possible by just warning the users right as they open it that they are missing dependencies.

> Can you let me know if you actually referenced any element from A to B? Or you just imported B into A and that's it? I will update the code to check package imports even if no reference is made to model B.

Yes, for example Model B might have an instantiated DDS4CCM template and Model A creates DDS_Listen and DDS_Write ports and connects them using a DDS Event.  Creating actual referenced elements didn't appear to change the behavior.

> Are these new models? or existing models that might have been opened in the workspace?

I took a new workspace and imported some of our example projects (BasicPubSub, ClientServer, IO_Connector). I filled out the Pathmap property for BasicPubSub and ClientServer. I had BasicPubSub import ClientServer and reference the interface "_obj" from ClientServer and then similarly did the same with IO_Connector importing BasicPubSub.

> I would suggest developing these ICM models as if developing a model library and at least restart the Papyrus once switched to the dynamic library. It is because that some other open models might have already loaded the model prior to switching to the pathmap. For example, if you open a selection dialog to set the type of CXAttribute then all models in the project or entire workspace are loaded depending on the search option even if you just cancel the dialog.
I did restart Papyrus without seeing the behavior that I would have expected. If I took a model without any pathmap specified and then specified a pathmap property, it was as if Papyrus didn't fully recognize it as dynamic and I had to delete that project and re-import it as an existing project in order for Papyrus to start warning about it being missing.


Even before we started this ticket, the pathmap feature worked great. All we are trying to do here is create as many safeguards as possible. If some of these are not possible, we can document them, but with our growing user base, any of these situations could cause a larger learning curve for our users.  Especially my first bullet of warning a user when they open model A that they are missing model B (if model B is indeed missing).  The second bullet isn't as important, because users won't be creating new ICMs all of the time, but I want to make sure that we aren't missing any situations where users could get tripped up with missing dependencies.

**Comment by ysroh** [comment]
Understood. I will let you know when I have the fix.

**Comment by ysroh** [comment]
@j26151 I merged the fix. I tested your scenario and added a part listener so we can check unresolved references. I still need to work on the validation issue so all broken links are detected during the validation.

**Comment by j26151** [comment]
@ysroh It appears that both issues that I brought up are fixed now. Can you elaborate on what you are still working on?

**Comment by ysroh** [comment]
@j26151 Can I close this issue then? The only other issue that we need to look into is the external reference validation issue.  I created a ticket for that. Issue #242.

**Comment by j26151** [comment]
@ysroh Yes, I wasn't sure what the issue was that you were still working on, but if there is another issue written for it then feel free to close this one.

**Comment by ysroh** [comment]
@j26151 Thanks. I will work on the validation issue on a separate ticket as it involves working with DDK.
